### ISSUE TITLE
Does it work with Mac

### ISSUE THREAD

**Issue by mallahyari** [gpt-omni/mini-omni2#22]
Hey there,

It seems it doesn't work with Mac as it gives this error: "AssertionError: Torch not compiled with CUDA enabled" when running this command: python3 server.py --ip '0.0.0.0' --port 60808

Is that working with Mac?

**Comment by YAY-3M-TA3** [comment]
No... whisper requires sparseMPS which is not implemented on torch.
 I did try, to use MLX-whisper instead...however:

Here is what I did - get the nightly torch - dont use 2.3.0:
```
pip install --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cpu
pip install mlx-whisper
```

then modified the requirements.txt and removed torch, torchvision, torchaudio and whisper before running: pip install -r requirements.txt

Then, before you can run the client Streamlit app on Mac, you need to add portAudio for arm64
`arch -arm64 /opt/homebrew/bin/brew install portaudio`

Then you can install pyaudio, but it needs to reference the arm64 portaudio
`pip3 install --no-cache-dir --global-option='build_ext' --global-option='-I/opt/homebrew/Cellar/portaudio/19.7.0/include' --global-option='-L/opt/homebrew/Cellar/portaudio/19.7.0/lib' PyAudio==0.2.14`

Next, there is an error in the mini-omni2 repo...  a missing file:
copy __init__.py from root into the /utils directory

Launch the server in a separate window (make sure to activate the conda environment)
`python3 server.py --ip '0.0.0.0' --port 60808 mps`
 
(client launch is:
`API_URL=http://0.0.0.0:60808/chat PYTHONPATH=$PYTHONPath=$PYTHONPATH:~/AI/mini-omni2 streamlit run webui/omni_streamlit.py`

replace ~/AI/mini-omni2 with your path to mini-omni2
)

At this point, the server will give errors during its warm-up phase...

modifications need to be made to inference.py and inference_vision.py to use mlx-whisper instead of whisper.
I downloaded https://huggingface.co/mlx-community/whisper-tiny-mlx-q4/tree/main to use and modified the whisper loader to get this model.

Modifications need to be made to **def get_input_ids_whisper_ATBatch(mel, leng, whispermodel, device):** to convert the mlx model to work with torch...
I made those changes, but I am stuck on this error:
```
/torch/nn/modules/linear.py", line 125, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: linear(): input and weight.T shapes cannot be multiplied (52x384 and 768x4864)
```

**Comment by mini-omni** [comment]
@mallahyari hi, you may change the device to cpu at https://github.com/gpt-omni/mini-omni2/blob/main/server.py#L11 to run with cpu.  For the mlx support, you may refer to: https://github.com/gpt-omni/mini-omni/issues/6

**Comment by YAY-3M-TA3** [comment]
Ok - I was able to get it running on Mac using Apple Silicon. (audio, no video.... yet....)

The main change from my previous post is to use this whisper model instead:
https://huggingface.co/mlx-community/whisper-small-mlx

you can download load it, then put it into the checkpoint directory.

Now the code changes:  (- = delete, + add this line)
**server.py**: line 17:  comment this out (this warms up the video, but, I haven't figure out the solution for the video MPS error )
`#self.client.warm_up()`


**inference.py**: 
line 25
```
- import whisper
+ import mlx_whisper as whisper
+ import mlx.core as mx
+ import mlx_whisper.audio as whisper_MLX_audio
+ import numpy as np
```
line 111
```
- mel = mel.unsqueeze(0).to(device)
+ dims = whispermodel.dims
+ mel = mx.array(np.random.randn(1, 3_000, dims.n_mels), mx.float16)
- audio_feature = whisper.embed_audio(mel)[0][:leng]
+ audio_feature = whispermodel.embed_audio(mel)[0][:leng]
```

```
- T = audio_feature.size(0)
+ T = audio_feature.shape[0]
```

line 145:
```
- return torch.stack([audio_feature, audio_feature]), stacked_inputids
+ return mx.stack([audio_feature, audio_feature]), stacked_inputids
```


line 149:
```
- audio = whisper_MLX_audio.load_audio(path)
+ audio = whisper_MLX_audio.load_audio(path)
    
- audio = whisper.pad_or_trim(audio)
+ audio = whisper_MLX_audio.pad_or_trim(audio)
- mel = whisper.log_mel_spectrogram(audio)
+ mel = whisper_MLX_audio.log_mel_spectrogram(audio)
```

line 363:
```
- whispermodel = whisper.load_model(whisper_model_path).to(device)
+ whispermodel = whisper.load_models.load_model(ckpt_dir, mx.float16)
```

line 447:
```
- audio_feature.to(torch.float32).to(model.device),
+ audio_feature,
```

**inference_vision.py**
line 13:
```
+ import mlx.core as mx
+ import numpy as np
```

line 33:
```
 - mel = mel.unsqueeze(0).to(device)
 + dims = whispermodel.dims
 + mel = mx.array(np.random.randn(1, 3_000, dims.n_mels), mx.float16)

- audio_len = audio_feature.size(0)
+ audio_len = audio_feature.shape[0]
```

line 69:
```
- return torch.stack([audio_feature,audio_feature]), stacked_inputids
+ return mx.stack([audio_feature, audio_feature]), stacked_inputids
```

line 141:
```
- audio_feature.to(torch.float32).to(self.device),
+ audio_feature,
```


**litgpt/model.py**
line 16:
`+ import numpy as np`

line 156:
```
- x_a = self.whisper_adapter(audio_features)
+ x_a = self.whisper_adapter(torch.from_numpy(np.array(audio_features)).to('mps'))
```


**Note:**  Since we commented out the "warm up", when you first run Streamlit and have it listen to your first voice command, it will return a "over flow error".  Just refresh your browser, then do it again and it should work from then on...

**Comment by YAY-3M-TA3** [comment]
Additionally, to solve for two errors with snac: 
```
return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: Placeholder storage has not been allocated on MPS device!
```

```
return F.conv1d(
NotImplementedError: Output channels > 65536 not supported at the MPS device. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.
```

**(export PYTORCH_ENABLE_MPS_FALLBACK=1 doesnt help )**

I made changes to the snac module to move it to cpu:
**snac/layers.py**: line 73
`+ self.model.to('cpu')`

**snac/vq.py**: line 94
```
- z_p_i = self.quantizers[i].decode_code(codes[i])
+ self.quantizers[i].to("cpu")
+ z_p_i = self.quantizers[i].decode_code(codes[i].to("cpu"))
```

There is still an issue when I run inference_vision.py to run the test, the resulting audio reponse is cut-off:
Here is the log:
```
using LLAMA MLP adapter for ASR feature
concat ImageQA_A feature
concat ImageQA_A feature
  0%|▍                                                                              | 9/1821 [00:00<02:30, 12.07it/s]
text output: The man appears to be in his middle years, judging
concat ImageQA_A feature
concat ImageQA_A feature
 16%|████████████▍                                                                | 295/1821 [00:22<01:55, 13.26it/s]
text output: The man appears to be in his middle years, judging by his mature features and the visible signs of aging such as wrinkles and a receding hairline. His complexion is fair, and he has a strong jawline and a confident, direct gaze. His eyes are blue, and he has a straight nose. He is wearing a dark suit with a white shirt and a dark tie. The background is neutral and blurred, focusing attention on the subject.
text_output: The man appears to be in his middle years, judging by his mature features and the visible signs of aging such as wrinkles and a receding hairline. His complexion is fair, and he has a strong jawline and a confident, direct gaze. His eyes are blue, and he has a straight nose. He is wearing a dark suit with a white shirt and a dark tie. The background is neutral and blurred, focusing attention on the subject.
```
The resulting data/samples/vision_qa_output.wav - is just 16KB and says, "The ma"

So, I dont know why this is getting cutoff... (there are no more errors)

**Comment by altunenes** [comment]
> Additionally, to solve for two errors with snac:
> 
> ```
> return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
> RuntimeError: Placeholder storage has not been allocated on MPS device!
> ```
> 
> ```
> return F.conv1d(
> NotImplementedError: Output channels > 65536 not supported at the MPS device. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.
> ```
> 
> **(export PYTORCH_ENABLE_MPS_FALLBACK=1 doesn't help )**
> 
> I made changes to the snac module to move it to cpu: **snac/layers.py**: line 73 `+ self.model.to('cpu')`
> 
> **snac/vq.py**: line 94
> 
> ```
> - z_p_i = self.quantizers[i].decode_code(codes[i])
> + self.quantizers[i].to("cpu")
> + z_p_i = self.quantizers[i].decode_code(codes[i].to("cpu"))
> ```
> 
> When I launch the server and launch the client streamlit app (in Chrome) with either Audio only or Audio+vision, **everything works as expected - (App can see and grab an image and the response is full and complete)**.
> 
> Anyways, hopefully this threads helps anyone else trying to run Mini-Omni2 on Mac Apple silicon. (M2 24Gb) It runs on the GPU and performs as good as what I see on the videos using the Nvidia hardware...

do you think 16gb m3 air can handle this model? I mean, for production use cases.
